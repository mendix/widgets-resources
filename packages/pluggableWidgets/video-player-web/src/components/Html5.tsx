import { Component, createElement, createRef } from "react";
import ReactResizeDetector from "react-resize-detector";

export interface Html5PlayerProps {
    url: string;
    poster?: string;
    autoPlay: boolean;
    showControls: boolean;
    loop: boolean;
    muted: boolean;
    style?: any;
    aspectRatio?: boolean;
}

export class Html5 extends Component<Html5PlayerProps> {
    private videoElement = createRef<HTMLVideoElement>();
    private errorElement = createRef<HTMLDivElement>();
    private readonly handleOnResize = this.onResize.bind(this);
    private readonly handleOnSuccess = this.handleSuccess.bind(this);
    private readonly handleOnError = this.handleError.bind(this);

    render(): JSX.Element {
        return (
            <div className="widget-video-player-html5-container">
                <div className="video-error-label-html5" ref={this.errorElement}>
                    The video failed to load :(
                </div>
                <video
                    className="widget-video-player-html5"
                    controls={this.props.showControls}
                    autoPlay={this.props.autoPlay}
                    muted={this.props.muted}
                    loop={this.props.loop}
                    poster={this.props.poster}
                    ref={this.videoElement}
                    height={!this.props.aspectRatio ? "100%" : undefined}
                    preload={this.props.poster ? "metadata" : "auto"}
                >
                    <source
                        src={this.props.url}
                        type="video/mp4"
                        onError={this.handleOnError}
                        onLoad={this.handleOnSuccess}
                    />
                    <ReactResizeDetector
                        handleWidth
                        handleHeight
                        onResize={this.handleOnResize}
                        refreshMode="debounce"
                        refreshRate={100}
                    />
                </video>
            </div>
        );
    }

    private handleError(): void {
        if (this.errorElement && this.errorElement.current) {
            this.errorElement.current.classList.add("hasError");
            if (this.videoElement && this.videoElement.current) {
                this.videoElement.current.controls = false;
            }
        }
    }

    private handleSuccess(): void {
        if (this.errorElement && this.errorElement.current) {
            this.errorElement.current.classList.remove("hasError");
            if (this.videoElement && this.videoElement.current) {
                this.videoElement.current.controls = this.props.showControls;
            }
        }
    }

    private onResize(): void {
        if (this.videoElement && this.videoElement.current && this.props.aspectRatio) {
            Html5.changeHeight(this.videoElement.current);
        }
    }

    private static changeHeight(element: HTMLElement): void {
        if (element.parentElement) {
            const height = element.clientHeight + "px";
            if (element.parentElement.parentElement) {
                element.parentElement.parentElement.style.height = height;

                if (element.parentElement.parentElement.parentElement) {
                    element.parentElement.parentElement.parentElement.style.height = height;
                }
            }
        }
    }
}
