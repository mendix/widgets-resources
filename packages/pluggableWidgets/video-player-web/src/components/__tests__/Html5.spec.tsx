import { createElement } from "react";

import { Html5, Html5PlayerProps } from "../Html5";
import { create } from "react-test-renderer";
import ReactResizeDetector from "react-resize-detector";

describe("Html5 Player", () => {
    const defaultProps = {
        url: "test",
        autoPlay: false,
        muted: false,
        loop: false,
        showControls: false,
        aspectRatio: false,
        poster: "test"
    };

    const defaulPlayer = (props: Html5PlayerProps): JSX.Element => <Html5 {...props} />;

    it("should renders correctly", () => {
        const player = create(defaulPlayer(defaultProps)).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with autoplay", () => {
        const player = create(defaulPlayer({ ...defaultProps, autoPlay: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with muted", () => {
        const player = create(defaulPlayer({ ...defaultProps, muted: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with controls", () => {
        const player = create(defaulPlayer({ ...defaultProps, showControls: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with loop", () => {
        const player = create(defaulPlayer({ ...defaultProps, loop: true })).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with poster", () => {
        const player = create(
            defaulPlayer({
                ...defaultProps,
                poster: "https://www.mendix.com/wp-content/themes/mendix/ui/images/homepage/air-status-app@2x.png"
            })
        ).toJSON();

        expect(player).toMatchSnapshot();
    });

    it("should renders correctly with aspectRatio", () => {
        const player = create(defaulPlayer({ ...defaultProps, aspectRatio: true }));
        window.dispatchEvent(new Event("resize"));
        const instance = player.root;
        const sizeDetector = instance.findByType(ReactResizeDetector);

        expect(player.toJSON).toMatchSnapshot();
        expect(sizeDetector).not.toBeNull();
        expect(sizeDetector.props).toHaveProperty("onResize");
    });
});
