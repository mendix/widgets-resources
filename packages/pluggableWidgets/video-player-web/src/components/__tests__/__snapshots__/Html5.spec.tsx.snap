// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Html5 Player should renders correctly 1`] = `
<div
  className="widget-video-player-html5-container"
>
  <div
    className="video-error-label-html5"
  >
    The video failed to load :(
  </div>
  <video
    autoPlay={false}
    className="widget-video-player-html5"
    controls={false}
    height="100%"
    loop={false}
    muted={false}
    poster="test"
    preload="metadata"
  >
    <source
      onError={[Function]}
      onLoad={[Function]}
      src="test"
      type="video/mp4"
    />
    <div />
  </video>
</div>
`;

exports[`Html5 Player should renders correctly with aspectRatio 1`] = `[Function]`;

exports[`Html5 Player should renders correctly with autoplay 1`] = `
<div
  className="widget-video-player-html5-container"
>
  <div
    className="video-error-label-html5"
  >
    The video failed to load :(
  </div>
  <video
    autoPlay={true}
    className="widget-video-player-html5"
    controls={false}
    height="100%"
    loop={false}
    muted={false}
    poster="test"
    preload="metadata"
  >
    <source
      onError={[Function]}
      onLoad={[Function]}
      src="test"
      type="video/mp4"
    />
    <div />
  </video>
</div>
`;

exports[`Html5 Player should renders correctly with controls 1`] = `
<div
  className="widget-video-player-html5-container"
>
  <div
    className="video-error-label-html5"
  >
    The video failed to load :(
  </div>
  <video
    autoPlay={false}
    className="widget-video-player-html5"
    controls={true}
    height="100%"
    loop={false}
    muted={false}
    poster="test"
    preload="metadata"
  >
    <source
      onError={[Function]}
      onLoad={[Function]}
      src="test"
      type="video/mp4"
    />
    <div />
  </video>
</div>
`;

exports[`Html5 Player should renders correctly with loop 1`] = `
<div
  className="widget-video-player-html5-container"
>
  <div
    className="video-error-label-html5"
  >
    The video failed to load :(
  </div>
  <video
    autoPlay={false}
    className="widget-video-player-html5"
    controls={false}
    height="100%"
    loop={true}
    muted={false}
    poster="test"
    preload="metadata"
  >
    <source
      onError={[Function]}
      onLoad={[Function]}
      src="test"
      type="video/mp4"
    />
    <div />
  </video>
</div>
`;

exports[`Html5 Player should renders correctly with muted 1`] = `
<div
  className="widget-video-player-html5-container"
>
  <div
    className="video-error-label-html5"
  >
    The video failed to load :(
  </div>
  <video
    autoPlay={false}
    className="widget-video-player-html5"
    controls={false}
    height="100%"
    loop={false}
    muted={true}
    poster="test"
    preload="metadata"
  >
    <source
      onError={[Function]}
      onLoad={[Function]}
      src="test"
      type="video/mp4"
    />
    <div />
  </video>
</div>
`;

exports[`Html5 Player should renders correctly with poster 1`] = `
<div
  className="widget-video-player-html5-container"
>
  <div
    className="video-error-label-html5"
  >
    The video failed to load :(
  </div>
  <video
    autoPlay={false}
    className="widget-video-player-html5"
    controls={false}
    height="100%"
    loop={false}
    muted={false}
    poster="https://www.mendix.com/wp-content/themes/mendix/ui/images/homepage/air-status-app@2x.png"
    preload="metadata"
  >
    <source
      onError={[Function]}
      onLoad={[Function]}
      src="test"
      type="video/mp4"
    />
    <div />
  </video>
</div>
`;
